plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -84,25 +100,145 @@ public DeleteModelTransportAction(
         Settings settings,
         NamedXContentRegistry xContentRegistry,
         ClusterService clusterService,
-        ModelAccessControlHelper modelAccessControlHelper
+        ModelAccessControlHelper modelAccessControlHelper,
+        RelatedModelIdHelper relatedModelIdHelper
     ) {
         super(MLModelDeleteAction.NAME, transportService, actionFilters, MLModelDeleteRequest::new);
         this.client = client;
         this.xContentRegistry = xContentRegistry;
         this.clusterService = clusterService;
         this.modelAccessControlHelper = modelAccessControlHelper;
+        this.relatedModelIdHelper = relatedModelIdHelper;
     }
 
     @Override
     protected void doExecute(Task task, ActionRequest request, ActionListener<DeleteResponse> actionListener) {
         MLModelDeleteRequest mlModelDeleteRequest = MLModelDeleteRequest.fromActionRequest(request);
         String modelId = mlModelDeleteRequest.getModelId();
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            // check whether agent are using them
+            SearchRequest searchAgentRequest = relatedModelIdHelper.constructQueryRequest(modelId);
+            client.search(searchAgentRequest, ActionListener.runBefore(ActionListener.wrap(searchResponse -> {
```

Comment:
Instead of querying agent, pipelines in a sequential manner, you can divide them to multiple single method responsible for single search and execute them parallel, each method can have a reference of CountDownLatch instance to coordinate them, refer: https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java#L243

Reply:
Make it parallel.